nx/lib/nx.ex

```diff
@@ -12609,26 +12702,78 @@ defmodule Nx do
         [1, 12, 112]
       >
 
+  ## Vectorized tensors
+
+  `tensor` and `indices` have their vectorized axes broadcast together,
+  and then the operation takes place normally, with `:axis` and `indices`
+  having their values in reference to the input shape.
+
+      iex> t = Nx.tensor([[[1, 2], [11, 12]], [[101, 102], [111, 112]]]) |> Nx.vectorize(:x)
+      iex> idx = Nx.tensor([[[0, 0], [0, 1]], [[1, 0], [1, 1]]]) |> Nx.vectorize(:x)
+      iex> Nx.gather(t, idx)
+      #Nx.Tensor<
+        vectorized[x: 2]
+        s64[2]
+        [
+          [1, 2],
+          [111, 112]
+        ]
+      >
+
+  And with vectorized broadcasting:
+
+      iex> t = Nx.tensor([[[1, 2], [11, 12]], [[101, 102], [111, 112]]]) |> Nx.vectorize(:x)
+      iex> idx = Nx.tensor([[[0, 0], [0, 1]], [[1, 0], [1, 1]]]) |> Nx.vectorize(:y)
+      iex> Nx.gather(t, idx)
+      #Nx.Tensor<
+        vectorized[x: 2][y: 2]
+        s64[2]
+        [
+          [
+            [1, 2],
+            [11, 12]
+          ],
+          [
+            [101, 102],
+            [111, 112]
+          ]
+        ]
+      >
+
   ## Error cases
 
       iex> Nx.gather(Nx.tensor([[1, 2], [3, 4]]), Nx.tensor([[0, 0]], type: :f32))
       ** (ArgumentError) indices must be an integer tensor, got {:f, 32}
   """
   @doc type: :indexed
   def gather(tensor, indices) do
-    tensor = to_tensor(tensor)
-    indices = to_tensor(indices)
-
-    Nx.Shared.raise_vectorized_not_implemented_yet(tensor, __ENV__.function)
-    Nx.Shared.raise_vectorized_not_implemented_yet(indices, __ENV__.function)
+    [%T{vectorized_axes: vectorized_axes} = tensor, indices] =
+      broadcast_vectors([tensor, indices])
 
     unless Nx.Type.integer?(indices.type) do
       raise ArgumentError, "indices must be an integer tensor, got #{inspect(indices.type)}"
     end
 
+    offset = length(vectorized_axes)
+
+    tensor = devectorize(tensor, keep_names: false)
+    indices = devectorize(indices, keep_names: false)
+
+    iota_shape =
+      indices.shape |> Tuple.delete_at(tuple_size(indices.shape) - 1) |> Tuple.append(1)
+
+    offset_axes = (offset - 1)..0//-1
+
+    indices =
+      offset_axes
+      |> Enum.reduce([indices], &[Nx.iota(iota_shape, axis: &1) | &2])
+      |> concatenate(axis: -1)
```

Comment:
Just to be safe, let's only do this if `offset != 0`, wdyt?

Reply:
Safe in terms of performance. I know concatenate will simply return itself and offset_axes is zero, but it is good to skip the extra calls anyway. :)